### PR TITLE
Add dict query support

### PR DESCRIPTION
**Pull request by clementetb**
This PR adds query functionality to dictionaries by adding the following methods to queries:

- `containsKey(fieldname, key)`: queries for dictionaries that contain a given key.
- `containsValue(fieldname, value)`: queries for dictionaries that contain a given value.
- `containsEntry(fieldname, entry)`: queries for dictionaries that contain an entry (pair key-value).

Entries are limited to keys of type String (dictionary use case). This limitation can be lifted once we support new key types.

### CODE REVIEW

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 1]
```diff
@@ -4118,6 +4128,72 @@ public void minimumDate_runOnMainThreadThrows() {
...
+        RealmResults<AllTypes> results = realm.where(AllTypes.class).containsKey(AllTypes.FIELD_STRING_DICTIONARY, "hello world1").findAll();
+        assertEquals(2, results.size());
+        assertEquals("Test1", results.get(0).getColumnStringDictionary().get("hello world1"));
+        assertEquals("Test2", results.get(1).getColumnStringDictionary().get("hello world1"));
+    }
+
+    @Test
+    public void dictionary_containsKeyNonLatin(){
+        fillDictionaryTests();
+        RealmResults<AllTypes> results = realm.where(AllTypes.class).containsKey(AllTypes.FIELD_STRING_DICTIONARY, "델타").findAll();
```
Comment (edualonso) [inline realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:4156]: Hm interesting. I haven't added tests for non-latin string keys and values in the dictionary tests. Maybe I should 🙊

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 2]
```diff
@@ -4118,6 +4128,72 @@ public void minimumDate_runOnMainThreadThrows() {
...
+        fillDictionaryTests();
+        RealmResults<AllTypes> results = realm.where(AllTypes.class).containsEntry(AllTypes.FIELD_STRING_DICTIONARY, new AbstractMap.SimpleImmutableEntry<>("hello world1", "Test2")).findAll();
+        assertEquals(1, results.size());
+        assertEquals("Test2", results.first().getColumnStringDictionary().get("hello world1"));
+    }
+
+    @Test
+    public void dictionary_containsKeyNull(){
+        fillDictionaryTests();
+        RealmResults<AllTypes> results = realm.where(AllTypes.class).containsKey(AllTypes.FIELD_STRING_DICTIONARY, null).findAll();
```
Comment (edualonso) [inline realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:4180]: Just FYI: dictionaries cannot contain null keys. We have a check on the Java side of things and throw an exception. I don't really know how it behaves when passing null keys to the OS, either for queries or writes. It seems it's accepted for queries from what I can see 🤔

**Review by edualonso — APPROVED** [review]
Looks terrific 🥇 I would suggest @cmelchior aka "the string-based query master" should have a look.

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java (hunk 1 above)
Comment (cmelchior) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:4156]: Probably a good idea, whitespace and non-lation characters tend to cause issues in funny places.

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 3]
```diff
@@ -4118,6 +4128,72 @@ public void minimumDate_runOnMainThreadThrows() {
...
+    public void dictionary_containsKeyNonLatin(){
+        fillDictionaryTests();
+        RealmResults<AllTypes> results = realm.where(AllTypes.class).containsKey(AllTypes.FIELD_STRING_DICTIONARY, "델타").findAll();
+        assertEquals(0, results.size());
+    }
+
+    @Test
+    public void dictionary_containsValue(){
+        fillDictionaryTests();
+        RealmResults<AllTypes> results = realm.where(AllTypes.class).containsValue(AllTypes.FIELD_STRING_DICTIONARY, "Test2").findAll();
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:4163]: Maybe also a test that doesn't match any values?

File: realm/realm-library/src/main/java/io/realm/Mixed.java [hunk 4]
```diff
@@ -455,6 +455,55 @@ public Decimal128 asDecimal128() {
...
+        } else if (RealmModel.class.isAssignableFrom(value.getClass())) {
+            RealmModel model = (RealmModel) value;
+
+            if (!RealmObject.isValid(model) || !RealmObject.isManaged(model)) {
+                throw new IllegalArgumentException("RealmObject is not a valid managed object.");
+            }
+
+            return Mixed.valueOf((RealmModel) model);
+        } else {
+            throw new IllegalArgumentException("Unsupported query argument type: " + value.getClass().getSimpleName());
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Mixed.java:504]: This error message doesn't look right? Shouldn't it just be "Type not supported Mixed: " ?

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 5]
```diff
@@ -1934,6 +1935,231 @@ public boolean isValid() {
         return this;
     }
 
+    /**
+     * Condition that value of field contains the specified key.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:1939]: We should make it clear this only works for dictionary types

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 6]
```diff
@@ -1934,6 +1935,231 @@ public boolean isValid() {
...
+
+    /**
+     * Condition that value of field contains the specified value.
+     *
+     * @param fieldName the field to compare.
+     * @param value     the value to search for.
+     * @return the query object.
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable Boolean value) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:1960]: We should make it clear this only works for Dictionaries (and I suspect we later want to use this for Sets as well?) Same for all the other methods

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 7]
```diff
@@ -1934,6 +1935,231 @@ public boolean isValid() {
...
+
+    /**
+     * Condition that value of field contains the specified value.
+     *
+     * @param fieldName the field to compare.
+     * @param value     the value to search for.
+     * @return the query object.
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, Mixed value) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2142]: There seems to be missing support for `containsValue(RealmModel)`?

File: realm/realm-library/src/main/java/io/realm/Mixed.java (hunk 4 above)
Comment (cmelchior) [reply realm/realm-library/src/main/java/io/realm/Mixed.java:504]: Ah wait, this is package protected. Nvm.

File: realm/realm-library/src/main/java/io/realm/internal/TableQuery.java [hunk 8]
```diff
@@ -344,6 +297,26 @@ public TableQuery containsInsensitive(String fieldName, Mixed value) {
         return this;
     }
 
+    // Dictionary queries
+
+    public TableQuery containsKey(String fieldName, Mixed key) {
+        mixedNativeFunctions.callRawPredicate(this, "ANY " + fieldName + ".@keys == $0", key);
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/TableQuery.java:303]: 🥇 🥳

**Review by cmelchior — CHANGES_REQUESTED** [review]
Nice 🎉  ... only minor thing is that we seem to be missing support for some types of input + a small clarification in the Javadoc.

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java (hunk 2 above)
Comment (clementetb) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:4180]: Queries allow null values anywhere, I am just testing that there is no edge case on dictionaries.

File: realm/realm-library/src/main/java/io/realm/Mixed.java (hunk 4 above)
Comment (clementetb) [reply realm/realm-library/src/main/java/io/realm/Mixed.java:504]: Copy-paste typo, it should be `Type not supported on Mixed: `. It is used internally but the exception can show to the user.

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 9]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable Short value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2004]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 10]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable Integer value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2018]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 11]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable Long value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2032]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 12]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable Double value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2046]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 13]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable Float value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2060]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 14]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable String value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2074]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 15]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable byte[] value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2088]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 16]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable Date value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2102]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 17]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable Decimal128 value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2116]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 18]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable ObjectId value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2130]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 19]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, @Nullable UUID value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2144]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 20]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, Mixed value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, value);
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified value.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2158]:
```suggestion
     * Condition that value, if dictionary field, contains the specified value.
```

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 21]
```diff
@@ -1943,6 +1944,245 @@ public boolean isValid() {
...
+     * @throws java.lang.IllegalArgumentException if the field name isn't valid.
+     */
+    public RealmQuery<E> containsValue(String fieldName, RealmModel value) {
+        realm.checkIfValid();
+        this.query.containsValue(realm.getSchema().getKeyPathMapping(), fieldName, Mixed.valueOf(value));
+        return this;
+    }
+
+    /**
+     * Condition that value if dictionary field contains the specified entry.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java:2172]:
```suggestion
     * Condition that value, if dictionary field, contains the specified entry.
```

**Review by cmelchior — APPROVED** [review]

**Pull request merged**